src/scipp/core/operations.py

```diff
@@ -292,3 +294,57 @@ def to(
         return to_unit(var.astype(dtype, copy=copy), unit=unit, copy=False)
     else:
         return to_unit(var, unit=unit, copy=copy).astype(dtype, copy=False)
+
+
+@overload
+def merge(lhs: Dataset, rhs: Dataset) -> Dataset:
+    ...
+
+
+@overload
+def merge(lhs: DataGroup, rhs: DataGroup) -> DataGroup:
+    ...
+
+
+def merge(lhs, rhs):
+    """Merge two datasets or data groups into one.
+
+    If an item appears in both inputs, it must have an identical value in both.
+
+    Parameters
+    ----------
+    lhs:
+        First dataset or data group.
+    rhs:
+        Second dataset or data group.
+
+    Returns
+    -------
+    :
+        A new object that contains the union of all data items,
+        coords, masks and attributes.
+
+    Raises
+    ------
+    scipp.DatasetError
+        If there are conflicting items with different content.
+    """
+    if isinstance(lhs, Dataset) or isinstance(rhs, Dataset):
+        return _call_cpp_func(_cpp.merge, lhs, rhs)
+    return _merge_data_group(lhs, rhs)
+
+
+def _generic_identical(a: Any, b: Any) -> bool:
+    try:
+        return identical(a, b)
+    except TypeError:
+        return a == b
```

Comment:
Since numpy is so common, maybe add another branch using `np.array_equal`?

Reply:
Good point!